packages/fetchai/connections/ledger_api/connection.py

```diff
@@ -0,0 +1,239 @@
+
+        envelope = Envelope(
+            to=self.address,
+            sender="",
+            protocol_id=message.protocol_id,
+            message=message,
+        )
+        return envelope
+
+
+class _RequestDispatcher:
+    def __init__(
+        self,
+        loop: Optional[asyncio.AbstractEventLoop],
+        executor: Optional[Executor] = None,
+    ):
+        """
+        Initialize the request dispatcher.
+
+        :param loop: the asyncio loop.
+        :param executor: an executor.
+        """
+        self.loop = loop if loop is not None else asyncio.get_event_loop()
+        self.executor = executor
+
+    async def run_async(self, func: Callable, *args):
+        """
+        Run a function in executor.
+
+        :param func: the function to execute.
+        :param args: the arguments to pass to the function.
+        :return: the return value of the function.
+        """
+        return await self.loop.run_in_executor(self.executor, func, *args)
+
+    def get_handler(
+        self, performative: LedgerApiMessage.Performative
+    ) -> Callable[[LedgerApi, LedgerApiMessage], Task]:
+        """
+        Get the handler method, given the message performative.
+
+        :param performative: the message performative.
+        :return: the method that will send the request.
+        """
+        handler = getattr(self, performative.value, lambda *args, **kwargs: None)
+        return handler
+
+    def dispatch(self, api: LedgerApi, message: LedgerApiMessage) -> Task:
+        """
+        Dispatch the request to the right sender handler.
+
+        :param api: the ledger api.
+        :param message: the request message.
+        :return: an awaitable.
+        """
+        performative = cast(LedgerApiMessage.Performative, message.performative)
+        handler = self.get_handler(performative)
+        return self.loop.create_task(self.run_async(handler, api, message))
+
+    def get_balance(self, api: LedgerApi, message: LedgerApiMessage):
+        """
+        Send the request 'get_balance'.
+
+        :param api: the API object.
+        :param message: the Ledger API message
+        :return: None
+        """
+        # TODO wrapping synchronous calls with multithreading to make it asynchronous.
+        #   LedgerApi async APIs would solve that.
+        balance = api.get_balance(message.address)
+        result = LedgerApiMessage(LedgerApiMessage.Performative.BALANCE, amount=balance)
+        return result
+
+    def get_transaction_receipt(self, api: LedgerApi, message: LedgerApiMessage):
+        """
+        Send the request 'get_transaction_receipt'.
+
+        :param api: the API object.
+        :param message: the Ledger API message
+        :return: None
+        """
+        # TODO what is a receipt? needs better data structures
+        # result = api.get_transaction_receipt(message.tx_digest)
+        return LedgerApiMessage(performative=LedgerApiMessage.Performative.TX_RECEIPT)
```

Comment:
Good question!

Reply:
For now; I'd say `Any` and then have a custom attribute in the protocol which we pickle (see search protocol for example). The reason being, it is too early to fix the data structure, we haven't seen many different ledgers yet.

Reply:
yes, done